### PR TITLE
Thinning

### PR DESCRIPTION
**Pull request by lucianolorenti**
Hello. I made a few changes to  #4 . I am using the implementation described in [Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.](https://pdfs.semanticscholar.org/6157/e6ec4fd66e66526fce9b207fb5e511b9f751.pdf)
I think it's the same one that matlab uses. However, I detected a few pixel differences between this implementation and Matlab in some images I tested. I will analyze that.

### CODE REVIEW

**Comment by juliohm** [comment]
Thank you @lucianolorenti and @vernwalrahul , I am at a conference this week, but will take a look at the PR and references that you have cited soon I am back next week.

Thanks for improving the previous proposed changes.

File: REQUIRE [hunk 1]
```diff
@@ -1,2 +1,3 @@
 julia 0.6
 ImageCore
+ImageFiltering
```
Comment (juliohm) [inline REQUIRE]: We can't afford ImageFiltering as a dependency, we had removed it in the past.

**Comment by juliohm** [comment]
I almost forgot this PR, should take a look at it this week.

File: REQUIRE (hunk 1 above)
Comment (lucianolorenti) [reply REQUIRE]: I didn't notice that ImageFiltering is broken in nightly. I will process the image borders by myself.

File: src/morphological_operations.jl [hunk 2]
```diff
@@ -99,3 +99,90 @@ which is defined as the arithmetic difference between the internal and the exter
...
+"""
+```
+function thinning_iteration(img::AbstractArray{Bool,2}; odd_iteration::Bool)
+```
+The thining iteration evaluates three conditions in order to determine which pixels of the image should be removed from img.
+The three conditions are explained in:
+Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+"""
+function thinning_iteration!(img_ori::AbstractArray{Bool,2}, odd_iteration::Bool)
+    local marker = trues(size(img_ori))
```
Comment (juliohm) [inline src/morphological_operations.jl]: I think we can remove all occurrences of the keyword `local` in the implementation. Did you choose to use `local` in these places for a reason?

File: src/morphological_operations.jl [hunk 3]
```diff
@@ -99,3 +99,90 @@ which is defined as the arithmetic difference between the internal and the exter
...
+            local B2 = (p2 || p3) + (p4 || p5) + (p6 || p7) + (p8 || p1)
+            local B  = min(B1,B2)
+
+            local G3 = false
+            if (odd_iteration)
+                G3 = (p2 || p3 || (!p5)) && p4                
+            else
+                G3 = (p6 || p7 || (!p1)) && p8
+            end
+            if (A == 1) && ((2<= B)  && (B <=3)) && (!G3)
```
Comment (juliohm) [inline src/morphological_operations.jl]: Please fix the white space, it is small detail but improves readability.

File: src/morphological_operations.jl [hunk 4]
```diff
@@ -99,3 +99,90 @@ which is defined as the arithmetic difference between the internal and the exter
...
+            local B  = min(B1,B2)
+
+            local G3 = false
+            if (odd_iteration)
+                G3 = (p2 || p3 || (!p5)) && p4                
+            else
+                G3 = (p6 || p7 || (!p1)) && p8
+            end
+            if (A == 1) && ((2<= B)  && (B <=3)) && (!G3) 
+                marker[i,j] =false
```
Comment (juliohm) [inline src/morphological_operations.jl]: Other white space detail

File: src/morphological_operations.jl [hunk 5]
```diff
@@ -99,3 +99,90 @@ which is defined as the arithmetic difference between the internal and the exter
...
+            end
+        end
+        
+    end
+    img_ori[:] =  img_ori[:].&(marker[:])
+end
```
Comment (juliohm) [inline src/morphological_operations.jl]: Please add a line break here to separate the doc string from the end of the function above.

File: src/morphological_operations.jl [hunk 6]
```diff
@@ -99,3 +99,90 @@ which is defined as the arithmetic difference between the internal and the exter
...
+```
+
+Applies a binary blob thinning operation, to achieve a skeletization of the input image.
+The procedure carries on until max_iterations is achieved or when there are no changes between two consecutives iterations.
+
+The algorithm is described in:
+1. Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+2. Lam, L., Lee, S. W., & Suen, C. Y. (1992). Thinning methodologies-a comprehensive survey. IEEE Transactions on pattern analysis and machine intelligence, 14(9), 869-885.
+"""
+function thinning(img::AbstractArray{T}; max_iterations::I=0) where T<:Bool where I<:Integer
```
Comment (juliohm) [inline src/morphological_operations.jl]: We need to generalize this function signature in the future, for now please change it to:

```julia
function thinning(img::AbstractArray{Bool}; maxiter::Integer)
end
```

File: src/morphological_operations.jl [hunk 7]
```diff
@@ -99,3 +99,90 @@ which is defined as the arithmetic difference between the internal and the exter
 `region` allows you to control the dimensions over which this operation is performed.
 """
 morpholaplace(img::AbstractArray, region=coords_spatial(img)) = dilate(img, region) + erode(img, region) - 2img
+
+
+struct FalsePadded
```
Comment (juliohm) [inline src/morphological_operations.jl]: By reading the paper, I learned that the algorithm only uses 3x3 neighorhoods. In that case, we can eliminate the class `FalsePadded`, create a larger image padded with zeros, and change the loop to `for i=2:h-1, j=2:w-1`. Makes sense?

File: src/morphological_operations.jl (hunk 2 above)
Comment (lucianolorenti) [reply src/morphological_operations.jl]: Yes. They can be removed.  maker is local no matter what, because of the assignment. I think I wrote it as an habit I should get rid off.

File: src/morphological_operations.jl (hunk 3 above)
Comment (lucianolorenti) [reply src/morphological_operations.jl]: You mean between the operators and the numbers?

File: src/morphological_operations.jl (hunk 7 above)
Comment (lucianolorenti) [reply src/morphological_operations.jl]: I was trying to avoid relocations. But, yes it is almost the same using the bigger image and assign 
    marker[i-1,j-1]=false 
to delete a pixel

File: src/morphological_operations.jl (hunk 2 above)
Comment (juliohm) [reply src/morphological_operations.jl]: Yes, `local` has a different purpose in the language, to preserve the value of an outside variable in a loop of the same name.

**Comment by juliohm** [comment]
Hi @lucianolorenti sorry for requesting so many changes, please understand that the projects under the Images.jl umbrella are used by many people, and we need to keep things as fast and as clean as possible.

**Comment by lucianolorenti** [comment]
Julio don't  worry. I understand what are you doing.I am grateful that you are taking time to review the code. I will have time tomorrow, I think, to view your new suggestions.

File: src/morphological_operations.jl [hunk 8]
```diff
@@ -99,3 +99,73 @@ which is defined as the arithmetic difference between the internal and the exter
...
+
+The algorithm is described in:
+1. Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+2. Lam, L., Lee, S. W., & Suen, C. Y. (1992). Thinning methodologies-a comprehensive survey. IEEE Transactions on pattern analysis and machine intelligence, 14(9), 869-885.
+"""
+function thinning(img::AbstractArray{Bool}; maxiter::Integer=0) 
+    prev = falses(size(img))
+    processed = copy(img)
+    it = 0
+    while ((maxiter==0) || (it<maxiter)) && (sum(prev.!=processed)>0)
```
[the remark this answers is not in the document]
Comment (lucianolorenti) [reply src/morphological_operations.jl]: Yes, the iterations are idempotent after a few steps. I wrote it not because a matter of convergence, but in case that is useful for someone not to perform the iterations until convergence. I remember I read some methods that used some threshold to stop the procedure. But sincecerly I couldn't find references about that. So, I suppose is not such a very common usecase.

[the remark this answers is not in the document]
Comment (juliohm) [reply src/morphological_operations.jl]: Makes sense, I thought you had this intention in mind :) Is there a situation where we want to thin an image to an intermediate (not fully thinned) state? I am just curious of the applications, it should be fine to keep the parameter there.

**Comment by juliohm** [comment]
It is starting to look very clean :ok_hand: and nice. I will give it another round of review when I find some time. It would be great to have the thinning function accepting an algorithm type to dispatch on, instead of constraining it to the specific implementation in that paper basically.

File: src/morphological_operations.jl [hunk 9]
```diff
@@ -99,3 +99,70 @@ which is defined as the arithmetic difference between the internal and the exter
...
+```
+
+Applies a binary blob thinning operation, to achieve a skeletization of the input image.
+The procedure carries on until there are no changes between two consecutives iterations.
+
+The algorithm is described in:
+1. Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+2. Lam, L., Lee, S. W., & Suen, C. Y. (1992). Thinning methodologies-a comprehensive survey. IEEE Transactions on pattern analysis and machine intelligence, 14(9), 869-885.
+"""
+function thinning(img::AbstractArray{Bool})
```
Comment (juliohm) [inline src/morphological_operations.jl]: It would be nice to allow future contributors to implement other thinning algorithms:

```julia
function thinning(img::AbstractArray{Bool}; algo::ThinAlgo=GuoAlgo())
    # dispatch appropriate implementation
    thinning_impl(img, algo)
end
```

The function would be dispatched on the algorithm type `GuoAlgo` or any other more informative name from the paper. The type could be empty like:

```julia
abstract type ThinAlgo end
struct GuoAlgo <: ThinAlgo end
```

since this specific algorithm doesn't have tuning parameters.

**Comment by lucianolorenti** [comment]
Julio. Sorry I didn't answer. I am  changing job and city. I will find time this week to check it out.

File: src/morphological_operations.jl (hunk 9 above)
Comment (lucianolorenti) [reply src/morphological_operations.jl]: I forgot to mention that after I added the code to process the image border I couldn't find differences with  matlab implementation anymore. I will add nevertheless the pages where the algorithm is explained.
I like the idea to have an algorithm type to dispatch on. But I have a question:: Where should I put the documentation of the method. ¿In the concrete type or in the function?

Comment (juliohm) [reply src/morphological_operations.jl]: That is awesome, good to know that the difference is gone. :) Maybe we should add the doc string to the type? What do you think? The general function could point to the individual doc strings like a "See also" section.

**Comment by juliohm** [comment]
Thank you @lucianolorenti for the heads up, no problem at all. We are all contributing in our free time :) I was also offline this weekend due to graduation and family trips. I am back tomorrow with more quick replies.

File: src/morphological_operations.jl [hunk 10]
```diff
@@ -99,3 +99,85 @@ which is defined as the arithmetic difference between the internal and the exter
...
+
+"""
+```
+function thinning_iteration(img::AbstractArray{Bool,2}; odd_iteration::Bool, algo::GuoAlgo)
+```
+The thining iteration evaluates three conditions in order to determine which pixels of the image should be removed from img.
+The three conditions are explained in the page 361 of:
+* Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+"""
+function thinning_iteration!(img_ori::AbstractArray{Bool,2}, odd_iteration::Bool, algo::GuoAlgo)
```
Comment (juliohm) [inline src/morphological_operations.jl]: The `algo` option is not needed in `thinning_iteration`, you can get rid of it.

The documentation is much nicer now, thanks for the doc strings!

File: src/morphological_operations.jl [hunk 11]
```diff
@@ -99,3 +99,85 @@ which is defined as the arithmetic difference between the internal and the exter
...
+    marker = trues(size(img_ori))
+    img = falses(size(img_ori).+2)
+    img[2:end-1,2:end-1] = img_ori
+    h, w = size(img)
+    for j=2:w-1, i=2:h-1
+        if !img[i,j]
```
Comment (juliohm) [inline src/morphological_operations.jl]: The following is better for code coverage and readability:

```julia
!img[i,j] && continue
```

File: src/morphological_operations.jl [hunk 12]
```diff
@@ -99,3 +99,85 @@ which is defined as the arithmetic difference between the internal and the exter
...
+"""
+```
+function thinning(img::AbstractArray{Bool}; algo::ThinAlgo=GuoAlgo())
+```
+Applies a binary blob thinning operation to achieve a skeletization of the input image.
+
+See also:
+* [`GuoAlgo`](@ref)
+"""
+function thinning(img::AbstractArray{Bool}; algo::ThinAlgo=GuoAlgo())
```
Comment (juliohm) [inline src/morphological_operations.jl]: Move this function up right after the type `GuoAlgo`. After that, you can move up the `thinning_impl`. The last function in the file should be `thinning_iteration`.

File: src/ImageMorphology.jl [hunk 13]
```diff
@@ -14,6 +14,8 @@ export
     tophat,
     bothat,
     morphogradient,
-    morpholaplace
-
+    morpholaplace,
```
Comment (juliohm) [inline src/ImageMorphology.jl]: Add a blank line after morpholaplace

File: src/morphological_operations.jl [hunk 14]
```diff
@@ -99,3 +99,85 @@ which is defined as the arithmetic difference between the internal and the exter
...
+        p4 = img[i,j+1]
+        p5 = img[i+1,j+1]
+        p6 = img[i+1,j]
+        p7 = img[i+1,j-1]
+        p8 = img[i,j-1]
+        A = (!p2 && (p3 || p4)) + (!p4 && (p5 || p6)) + (!p6 && (p7 || p8)) + (!p8 && (p1 || p2))
+        B1 = (p1 || p2) + (p3 || p4) + (p5 || p6) + (p7 || p8)
+        B2 = (p2 || p3) + (p4 || p5) + (p6 || p7) + (p8 || p1)
+        B = min(B1, B2)
+        G3 = false
```
Comment (juliohm) [inline src/morphological_operations.jl]: You can remove this line. "if" statements in Julia don't create namespaces, and the variable `G3` inside of the `if-else` will be visible outside.

File: src/morphological_operations.jl [hunk 15]
```diff
@@ -99,3 +99,85 @@ which is defined as the arithmetic difference between the internal and the exter
...
+        p1 = img[i-1,j-1]
+        p2 = img[i-1,j]
+        p3 = img[i-1,j+1]
+        p4 = img[i,j+1]
+        p5 = img[i+1,j+1]
+        p6 = img[i+1,j]
+        p7 = img[i+1,j-1]
+        p8 = img[i,j-1]
+        A = (!p2 && (p3 || p4)) + (!p4 && (p5 || p6)) + (!p6 && (p7 || p8)) + (!p8 && (p1 || p2))
+        B1 = (p1 || p2) + (p3 || p4) + (p5 || p6) + (p7 || p8)
```
Comment (juliohm) [inline src/morphological_operations.jl]: Consider renaming this variables to their names in the paper `N1`, `N2`, and `N`. Any other variable that you can map directly to the name in the paper will help other people reading and contributing to the code in the future.

File: src/morphological_operations.jl [hunk 16]
```diff
@@ -99,3 +99,85 @@ which is defined as the arithmetic difference between the internal and the exter
...
+        B2 = (p2 || p3) + (p4 || p5) + (p6 || p7) + (p8 || p1)
+        B = min(B1, B2)
+        G3 = false
+        if odd_iteration
+            G3 = (p2 || p3 || (!p5)) && p4
+        else
+            G3 = (p6 || p7 || (!p1)) && p8
+        end
+        if (A == 1) && ((2 <= B) && (B <= 3)) && (!G3)
```
Comment (juliohm) [inline src/morphological_operations.jl]: After you rename `A` to `C`, `B` to `N`, and maybe `G3` to `O` like in the paper, this line can be written as:

```julia
if C == 1 && (2 ≤ N ≤ 3) && !O
```

Even better would be to save calculation by identifying which of these three conditions is more expensive to compute, and only compute it if the other conditions were already satisfied.

File: src/morphological_operations.jl [hunk 17]
```diff
@@ -99,3 +99,85 @@ which is defined as the arithmetic difference between the internal and the exter
...
+"""
+```
+function thinning_iteration(img::AbstractArray{Bool,2}; odd_iteration::Bool, algo::GuoAlgo)
+```
+The thining iteration evaluates three conditions in order to determine which pixels of the image should be removed from img.
+The three conditions are explained in the page 361 of:
+* Guo, Z., & Hall, R. W. (1989). Parallel thinning with two-subiteration algorithms. Communications of the ACM, 32(3), 359-373.
+"""
+function thinning_iteration!(img_ori::AbstractArray{Bool,2}, odd_iteration::Bool, algo::GuoAlgo)
+    marker = trues(size(img_ori))
```
Comment (juliohm) [inline src/morphological_operations.jl]: The variable `marker` is unnecessary, we should be deactivating pixels directly in `img_ori`.

**Comment by juliohm** [comment]
Hi @lucianolorenti are you working on this? I can take care of the final details if you are too busy.

**Comment by lucianolorenti** [comment]
Julio. Hello!. Yes I am still working on this.These weeks have been a bit busier than usual. 
I changed what you suggested. Is the functions order ok?
I wanted to do what you mention about  identifying which condition is more expensive. Should I do it with @profile ? Or analyzing which line does  more comparisons?

**Comment by juliohm** [comment]
Thank you @lucianolorenti , it is looking :ok_hand: Though I think there is a small issue with the logic in `thinning_impl` where you start with `prev=falses`, shouldn't it start with the original image instead?

After you review the logic in `thinning_impl`, potentially adding a new test for the corner case where the input image is already thinned, we should be ready to merge. Could you please move all the thinning-related functions you contributed to a separate file in the source called `thinning.jl`? I think the operation deserves its own place in the package for future contributions and review.

We could discuss optimizations in a separate PR, I see some opportunities there. What do you think? Let's get a working version merged first, so that users can benefit :slightly_smiling_face:

**Comment by juliohm** [comment]
@lucianolorenti could you please submit a new PR from the latest master branch? I have updated the package for Julia v0.7, and this has introduced conflicts with the changes here.

Please submit a new PR with a separate file called `thinning.jl` where you copy/paste the current implementation you have. Also, please copy the tests you have written to the appropriate section in the new runtests.jl file.

Thanks,

**Comment by lucianolorenti** [comment]
Julio I added the test and separated the implementation to a new file.

About what you wrote:

> Though I think there is a small issue with the logic in thinning_impl where you start with prev=falses, shouldn't it start with the original image instead?

If `prev` starts with the original image the while condition `prev != processed` will be false. I initialized it in that way to ensure that at least one time `thinning_iteration! `is executed. Perhaps I can write
`while it == 0 || prev != processed` and leave `prev` uninitialized.  Or using a boolean variable and update it inside the while.

I am going to make the new PR now.

**Comment by juliohm** [comment]
I will think about the logic again when you create the new PR. Thank you for the heads up.

**Comment by juliohm** [comment]
Superseeded by #7
